File: 07-trees/01448-Count_Good_Nodes_In_Binary_Tree/01448-count_good_nodes_in_binary_tree.cpp
// 1448: Count Good Nodes In Binary Tree
// https://leetcode.com/problems/count-good-nodes-in-binary-tree/


#include <iostream>
#include <queue>
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

TreeNode* creator(vector<int> values, TreeNode** root, int i, int n) {
    if (n==0) return NULL;
    if (i<n) {
        TreeNode *temp = new TreeNode(values[i]);
        *root = temp;
        (*root)->left = creator(values, &((*root)->left), 2*i+1, n);
        (*root)->right = creator(values, &((*root)->right), 2*i+2, n);
    }
    return *root;
};

void createTree(TreeNode** root, vector<int> inputs) {
    creator(inputs, root, 0, inputs.size());
}

void showTree(TreeNode* root) {
    queue<TreeNode*> q;
    vector<vector<int>> result = {};
    vector<int> c;
    if (root==NULL) { cout << "Empty !" << endl; return; }
    q.push(root);
    q.push(NULL);
    while (!q.empty()) {
        TreeNode *t = q.front();
        q.pop();
        if (t==NULL) {
            result.push_back(c);
            c.resize(0);
            if (q.size() > 0) q.push(NULL);
        } else {
            c.push_back(t->val);
            if (t->left) q.push(t->left);
            if (t->right) q.push(t->right);
        }
    }

    cout<<"["; for (auto x : result) {
        cout<<"["; for (auto y : x) {
            if (!y) { cout << "NULL,"; continue; }
            cout<<y<<",";
        } cout<<"\b],";
    } cout<<"\b]"<<endl;
}



class Solution {
private:
    int traverse(TreeNode *root, int max) {
        if(root!=NULL && root->val!=NULL && root->val>max)
            max = root->val;
        return root!=NULL && root->val!=NULL ? (root->val == max) + traverse(root->left, max) + traverse(root->right, max) : 0;
    }

public:
    // SOLUTION
    int goodNodes(TreeNode* root) {
        return traverse(root, -10000);
    }
};



int main() {
    Solution o;
    TreeNode* root = NULL;

    // INPUT
    vector<int> tn = {3,1,4,3,NULL,1,5};
    createTree(&root, tn);

    // OUTPUT
    auto result = o.goodNodes(root);
    cout << result << endl;

    return 0;
}

File: 07-trees/01448-Count_Good_Nodes_In_Binary_Tree/01448-count_good_nodes_in_binary_tree.py
# 1448: Count Good Nodes In Binary Tree
# https://leetcode.com/problems/count-good-nodes-in-binary-tree/



class TreeNode:
    def __init__(self, x: int):
        self.val = x
        self.left = None
        self.right = None

class Tree:
    def __init__(self) -> None:
        self.root = None

    def creator(self, values: list[int], root: TreeNode, i: int, n: int) -> TreeNode:
        if n==0 : return None
        if i<n:
            temp = TreeNode(values[i])
            root = temp
            root.left = self.creator(values, root.left, 2*i+1, n)
            root.right = self.creator(values, root.right, 2*i+2, n)
        return root

    def createTree(self, inputs: list[int]) -> None:
        self.root = self.creator(inputs, self.root, 0, len(inputs))
    
    def showTree(self, root: TreeNode) -> None:
        q = []
        result = []
        c = []
        if root==None : return result
        q.append(root)
        q.append(None)
        while len(q)!=0:
            t = q.pop(0)
            if t==None:
                result.append(c)
                c = []
                if len(q) > 0 : q.append(None)
            else:
                c.append(t.val)
                if t.left!=None : q.append(t.left)
                if t.right!=None : q.append(t.right)

        print("[", end="")
        for x in result:
            if len(x)==0 : continue
            print("[", end="")
            for y in x:
                if y==None: 
                    print("NULL,", end="")
                    continue
                print(y, end=",")
            print("\b]", end=",")
        print("\b]")


class Solution:
    # SOLUTION
    def goodNodes(self, root: TreeNode) -> int:        
        def traverse(root: TreeNode, max: int) -> None:
            if root!=None and root.val!=None and root.val>max:
                max = root.val
            return (root.val == max) + traverse(root.left, max) + traverse(root.right, max) if root!=None and root.val!=None else 0

        return traverse(root, -10000)


if __name__ == "__main__":
    o = Solution()
    tree = Tree()

    # INPUT
    tn = [3,1,4,3,None,1,5]
    tree.createTree(tn)

    # OUTPUT
    result = o.goodNodes(tree.root)
    print(result)
